web/packages/teleport/src/services/resources/types.ts

```diff
@@ -63,6 +63,29 @@ type RoleConditions = {
   node_labels?: Record<string, string[]>;
 };
 
+export type RoleOptions = {
+  cert_format: string;
+  create_db_user: boolean;
+  create_desktop_user: boolean;
+  desktop_clipboard: boolean;
+  desktop_directory_sharing: boolean;
+  enhanced_recording: string[];
+  forward_agent: boolean;
+  idp: {
+    saml?: {
```

Comment:
why is `saml` optional?

Reply:
This type describes the `Role` as it's seen by the frontend, not as it's described on the backend. There's a very subtle quirk there that stems from the `omitempty` attribute and from how our `Rolev6.CheckAndSetDefaults` work: if you ask Teleport to create a resource with `idp` field set to `null`, it's instead going to create the entire `idp->saml->enabled` structure. However, it's possible to create a role with `idp` set to an empty object, and the server will retain this state. So effectively, why the default state of `idp` itself is `{saml:{enabled:true}}`, the default value of the `saml` itself is `null`. This makes it effectively an optional field, and this is what I'm reflecting here. I'll put it in a comment.